lib/common/grpc-service.js

```diff
@@ -0,0 +1,354 @@
+var path = require('path');
+var snakeize = require('snakeize');
+
+/**
+ * @type {module:common/service}
+ * @private
+ */
+var Service = require('./service.js');
+
+/**
+ * @const {object} - A map of protobuf codes to HTTP status codes.
+ * @private
+ */
+var HTTP_ERROR_CODE_MAP = {
+  0: {
+    code: 200,
+    message: 'OK'
+  },
+
+  1: {
+    code: 499,
+    message: 'Client Closed Request'
+  },
+
+  2: {
+    code: 500,
+    message: 'Internal Server Error'
+  },
+
+  3: {
+    code: 400,
+    message: 'Bad Request'
+  },
+
+  4: {
+    code: 504,
+    message: 'Gateway Timeout'
+  },
+
+  5: {
+    code: 404,
+    message: 'Not Found'
+  },
+
+  6: {
+    code: 409,
+    message: 'Conflict'
+  },
+
+  7: {
+    code: 403,
+    message: 'Forbidden'
+  },
+
+  8: {
+    code: 429,
+    message: 'Too Many Requests'
+  },
+
+  9: {
+    code: 412,
+    message: 'Precondition Failed'
+  },
+
+  10: {
+    code: 409,
+    message: 'Conflict'
+  },
+
+  11: {
+    code: 400,
+    message: 'Bad Request'
+  },
+
+  12: {
+    code: 501,
+    message: 'Not Implemented'
+  },
+
+  13: {
+    code: 500,
+    message: 'Internal Server Error'
+  },
+
+  14: {
+    code: 503,
+    message: 'Service Unavailable'
+  },
+
+  15: {
+    code: 500,
+    message: 'Internal Server Error'
+  },
+
+  16: {
+    code: 401,
+    message: 'Unauthorized'
+  }
+};
+
+/**
+ * Service is a base class, meant to be inherited from by a "service," like
+ * BigQuery or Storage.
+ *
+ * This handles making authenticated requests by exposing a `makeReq_` function.
+ *
+ * @param {object} config - Configuration object.
+ * @param {string} config.baseUrl - The base URL to make API requests to.
+ * @param {string[]} config.scopes - The scopes required for the request.
+ * @param {object} options - [Configuration object](#/docs/?method=gcloud).
+ */
+function GrpcService(config, options) {
+  Service.call(this, config, options);
+
+  var service = config.service;
+  var apiVersion = config.apiVersion;
+  var rootDir = googleProtoFiles('..');
+
+  if (config.customEndpoint) {
+    this.grpcCredentials = grpc.credentials.createInsecure();
+  }
+
+  this.protoOpts = config.proto;
+  this.proto = grpc.load({
+    root: rootDir,
+    file: path.relative(rootDir, googleProtoFiles[service][apiVersion])
+  }).google[service][apiVersion];
+}
+
+nodeutil.inherits(GrpcService, Service);
+
+/**
+ * Make an authenticated request with gRPC.
+ *
+ * @param {object} protoOpts - The proto options.
+ * @param {string} protoOpts.service - The service name.
+ * @param {string} protoOpts.method - The method name.
+ * @param {number=} protoOpts.timeout - After how many milliseconds should the
+ *     request cancel.
+ * @param {object} reqOpts - The request options.
+ * @param {function=} callback - The callback function.
+ */
+GrpcService.prototype.request = function(protoOpts, reqOpts, callback) {
+  if (global.GCLOUD_TEST_ENVIRONMENT) {
+    return global.GCLOUD_TEST_ENVIRONMENT;
```

Comment:
Hmm, we probably need a better name for that. It is strictly there to prevent the docs test file from making requests. `MITM` is used to prevent HTTP requests going through, but this doesn't stop gRPC requests. I tried to stub gRPC with mockery and any other kind of hackery I could think of, but this was the only thing that worked.

Can you think of a better name, like `GCLOUD_DOCS_TEST_SCRIPT` for maximum literacy?